tiatoolbox/models/controller/semantic_segmentor.py

```diff
@@ -0,0 +1,1017 @@
+
+    """
+
+    def __init__(
+        self,
+        ioconfig: IOSegmentorConfig,
+        wsi_paths: List[Union[str, pathlib.Path]],
+        mp_shared_space: Namespace,
+        preproc: Callable[[np.ndarray], np.ndarray] = None,
+        mode="wsi",
+    ):
+        super().__init__()
+        self.mode = mode
+        self.preproc = preproc
+        self.ioconfig = copy.deepcopy(ioconfig)
+
+        if mode == "tile":
+            warnings.warn(
+                " ".join(
+                    [
+                        "WSIPatchDataset only reads image tile at",
+                        '`units="baseline"`. Resolutions will be converted',
+                        "to baseline value.",
+                    ]
+                )
+            )
+            self.ioconfig = self.ioconfig.to_baseline()
+
+        self.mp_shared_space = mp_shared_space
+        self.wsi_paths = wsi_paths
+        self.wsi_idx = None  # to be received externally via thread communication
+        self.reader = None
+
+    def _get_reader(self, img_path):
+        """Get approriate reader for input path."""
+        img_path = pathlib.Path(img_path)
+        if self.mode == "wsi":
+            reader = get_wsireader(img_path)
+        else:
+            img = imread(img_path)
+            # initialise metadata for VirtualWSIReader.
+            # here, we simulate a whole-slide image, but with a single level.
+            metadata = WSIMeta(
+                mpp=np.array([1.0, 1.0]),
+                objective_power=10,
+                slide_dimensions=np.array(img.shape[:2][::-1]),
+                level_downsamples=[1.0],
+                level_dimensions=[np.array(img.shape[:2][::-1])],
+            )
+            reader = VirtualWSIReader(
+                img,
+                metadata,
+            )
+        return reader
+
+    def __len__(self):
+        return len(self.mp_shared_space.patch_inputs)
+
+    @staticmethod
+    def collate_fn(batch):
+        """Prototype to handle reading exception.
+
+        This will exclude any sample with `None` out from the
+        batch. As such, wrapping `__getitem__` with try-catch
+        and return `None` upon exceptions will prevent crashing the
+        entire programs. But as side effect, batch may not have size
+        as defined.
+
+        """
+        batch = [v for v in batch if v is not None]
+        return torch.utils.data.dataloader.default_collate(batch)
+
+    def __getitem__(self, idx: int):
+        # ! no need to lock as we do not modify source value in shared space
+        if self.wsi_idx != self.mp_shared_space.wsi_idx:
+            self.wsi_idx = int(self.mp_shared_space.wsi_idx.item())
+            self.reader = self._get_reader(self.wsi_paths[self.wsi_idx])
+
+        # this is in XY and at requested resolution not baseline
+        bound = self.mp_shared_space.patch_inputs[idx]
+        bound = bound.numpy()  # expected to be torch.Tensor
```

Comment:
Is this a single bound or a set of bounds. It looks like it should be plural to me.